field/TinyLang/Field/Evaluator.hs

```diff
@@ -144,57 +190,70 @@ evalBinOp Mul x y = return . UniConst Field $ x `mul` y
 evalBinOp Div x y = UniConst Field <$> divEval x y
 evalBinOp BAt x y = asIntegerEval x >>= fmap (UniConst Bool) . flip atEval y . fromIntegral
 
-evalStatementUni
-    :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
-    => Env (SomeUniConst f) -> Statement f -> Expr f a -> m (UniConst f a)
-evalStatementUni env (ELet (UniVar _ var) def) rest = do
-    defR <- evalExprUni env def
-    evalExprUni (insertVar var (Some defR) env) rest
-evalStatementUni env (EAssert expr) rest = do
-    exprR <- evalExpr env expr
-    if exprR
-        then evalExprUni env rest
-        else throwError $ AssertionFailedEvalError expr
+-- | Evaluate a program
+evalProgram :: (Monad m, Eq f, Field f, AsInteger f)
+    => Program f -> EvalT f m ()
+evalProgram = flip evalStatements (pure ()) . unProgram
+
+-- TODO:  Verify whether it is acutally right fold
+evalStatements :: (Monad m, Eq f, Field f, AsInteger f)
+    => Statements f -> EvalT f m () -> EvalT f m ()
+evalStatements (Statements stmts) kont = foldr evalStatement kont stmts
+
+-- | Evaluate a single statement that continue as @kont@
+evalStatement :: (Monad m, Eq f, Field f, AsInteger f)
+    => Statement f -> EvalT f m () -> EvalT f m ()
+evalStatement (ELet (UniVar _ var) def) kont = do
+    defR <- evalExprUni def
+    let someDefR = Some defR
+    modify (someDefR:)
+    withVar var someDefR kont
+evalStatement (EAssert expr) kont = do
+    exprR <- evalExpr expr
+    unless exprR $ throwError $ AssertionFailedEvalError expr
+    kont
+evalStatement (EFor (UniVar _ var) start end body) kont
+    | start <= end = foldr go kont [start..end]
+    | otherwise    = withVar var actualEnd kont
+        where
+            go i fkont = withVar var (Some $ fromIntegral i) $
+                            evalStatements body fkont
+            actualEnd = Some $ fromIntegral $ max start end
```

Comment:
As per previous discussion we already know in the `otherwise` clause that `start > end` and hence can do `withVar var start kont` there instead, but not a big deal.